File: solutions/erlang/bob/2/src/bob.erl
-module(bob).

-export([response/1]).

response(String) ->
    Trimmed = string:trim(String),
    Is_empty = Trimmed == [],
    Is_yell = is_yell(Trimmed),
    Is_question = is_question(Trimmed),
    if
        Is_empty -> "Fine. Be that way!";
        Is_yell and Is_question -> "Calm down, I know what I'm doing!"; 
        Is_yell -> "Whoa, chill out!";
        Is_question -> "Sure.";
        true -> "Whatever."
    end.

is_yell(String) -> 
    String =:= string:to_upper(String) andalso
    String =/= string:to_lower(String).

is_question([]) -> false;
is_question(_String) ->
    $? == lists:last(_String).

File: solutions/go/bob/1/bob.go
// Package bob provides a conversation library
package bob

import "strings"

// Hey makes it possible to communicate with Bob
func Hey(remark string) string {
	remark = strings.TrimSpace(remark)
	switch {
	case 0 == len(remark):
		return "Fine. Be that way!"
	case isQuestion(remark) && isYell(remark):
		return "Calm down, I know what I'm doing!"
	case isQuestion(remark):
		return "Sure."
	case isYell(remark):
		return "Whoa, chill out!"
	default:
		return "Whatever."
	}
}

func isQuestion(text string) bool {
	return text[len(text)-1:] == "?"
}

func isYell(text string) bool {
	return text == strings.ToUpper(text) && strings.ToLower(text) != text
}